### ISSUE TITLE
Trouble executing Python Script - RuntimeError: cannot unmarshal code objects in restricted execution mode (Ubuntu 16.04 LTS)

### ISSUE THREAD

**Issue by peos3** [mrj0/jep#62]
I have a Python script working with a [spacy](https://spacy.io) module.
The project runs fine on Windows 10 64bit, but when executing on Ubuntu 16.04 LTS I get this error message:
`jep.JepException: <type 'exceptions.RuntimeError'>: cannot unmarshal code objects in restricted execution mode
	at /home/pascal/.local/lib/python2.7/site-packages/pathlib.open(pathlib.py:1077)
	at spacy/syntax/parser.spacy.syntax.parser.Parser.load (spacy/syntax/parser.cpp:5023)(parser.pyx:92)
	at /home/pascal/.local/lib/python2.7/site-packages/spacy/language.create_parser(language.py:102)
	at /home/pascal/.local/lib/python2.7/site-packages/spacy/language.__init__(language.py:274)
	at /home/pascal/.local/lib/python2.7/site-packages/spacy/__init__.load(__init__.py:47)`

I have no clue how to fix this.
After reading some time [on the internet](http://stackoverflow.com/questions/40443573/python-constructor-not-accessible-in-restricted-mode), I found someone that says the following:

> Only know that error as being able to be possible to trigger when you are using Python sub interpreters via the C API and you are incorrectly sharing Python objects across interpreter contexts. Are you implementing your module as a C extension? Are you running it in a system which uses sub interpreters, such as mod_wsgi or other embedded system?

Maybe someone knows how to fix this or where to look: on the jeb side or the python environment side or even the spacy side.
Also that it is completely working fine on Windows is interesting.

Thanks in advance everyone for looking into this problem.

**Comment by ndjensen** [comment]
Jep is using sub-interpreters but we're not sharing across interpreter contexts unless you use the [shared module](https://github.com/mrj0/jep/blob/master/release_notes/3.6-notes.rst#python-shared-modules-beta) feature.  You might try using that feature, adding spacy as a shared module, and see if that resolves the issue.

[restricted mode](https://docs.python.org/2/library/restricted.html) seems to be disabled by default.  When you did your tests on Windows were you still using Python 2 or were you using Python 3?  Restricted mode seems to be gone in Python 3.

See example code on this [Stack Overflow post](http://stackoverflow.com/questions/1994117/what-is-pythons-restricted-execution-mode).  Can you try running python at the command line like the example and see if the f.func_globals() works?  Can you also try running the jep command line and see if the f.func_globals() works?

**Comment by jvstein** [comment]
Are you loading the spacy parser more than once in a single python interpreter (e.g. by evaling your script more than once)? My code creates a single spacy parser in a single Jep interpreter and re-uses it for the entire life-cycle of the application and it's been working well.

You might also consider compiling python manually on Ubuntu. I've run into a handful of problems with embedding due to their packaging and customizations.

**Comment by peos3** [comment]
We added spacy as a shared module to jep. We also used Python 2.7 and you are right, switching to Python 3.5 solved the restricted mode problem.

However, after rewriting our script for Python 3.5, we got a Java Virtual Machine Fatal Error (segmentation fault), the problemativ frame was "libpython3.5.so.1.0".

After fiddeling around i found the problem:
We used the getValue function like this:
`jep.getValuel("tagger.getTags(text,ents)");`  
So we directly called the Python function. If i rewrite the code like this, everything works fine:  
```{engine='python'}
jep.eval("result = tagger.getTags(text,ents)");
jep.getValue("result");
```

I don't know why this works though.

**Comment by ndjensen** [comment]
Thanks.  What version of Jep are you using?

**Comment by bsteffensmeier** [comment]
I might have the problem figured out, I was able to cause a crash on python 2.7 but I could not crash on python 3.6 using the following code:

```java
    public static void main(String[] args) throws JepException, IOException {
        Jep jep = new Jep(new JepConfig());
        jep.eval("import numpy");
        jep.getValue("numpy.ndarray([1,2])");
        jep.close();
    }

```

My theory is that [this line](https://github.com/mrj0/jep/blob/6cad5feffeeb0b39a9800ffca3df33b72767018a/src/jep/pyembed.c#L1188) can cause a crash since we DECREF the result after releasing the python thread. As long as the DECREF does not cause the object to be deallocated then no errors occur but if it does cause deallocation then arbitrary code may be executed that is expecting the python thread state to be active, if any of this code tries to manipulate the python thread state or the GIL then it will result in a crash.

Based off this theory, the reason using result fixes the problem is because the result variable will still hold a reference to the object that is being returned so it cannot be deallocated.

To replicate the problem I needed an object that would mess with thread state during deallocation, I tried lists, maps, tuples, ints, strings and they all work just fine but an ndarray failed on my python 2.7. Im not sure why it is not crashing on 3.6 but I have a much newer numpy. I don't think the problem has anything to do with numpy so I need to investigate more to find an object with a complex deallocation function that I can use in a unit test.

I've talked to @ndjensen and since this is a stability problem we would like to fix this and then release jep 3.6.3. Before I fix it I would like to find a more reliable test case and also verify that this is the same problem that you are seeing.

@peos3 can you provide more information about what result is, it looks like it may be a list or dict, if so what type() of objects does it hold? Also it would be helpful if you can verify the fix, just move the Py_XDECREF above the PyEval_ReleaseThread and rebuild and the problem should go away.

**Comment by bsteffensmeier** [comment]
The fix has been checked into the dev_3.6 branch and there is also unit tests for this issue in the dev_3.7 branch,

**Comment by ndjensen** [comment]
I'm going to close this since @bsteffensmeier fixed the stability issue and restricted mode is deprecated in Python.  The stability fix is included in Jep 3.6.3.
